File: workspace/judoku/src/main/java/judoku/solvers/IfValueFixedToLineOrColumnInsideABlockThenDeleteFromRest.java
package judoku.solvers;

import java.util.List;

import judoku.data.Block;
import judoku.data.Cell;
import judoku.data.Grid;
import judoku.data.OneToNine;

public class IfValueFixedToLineOrColumnInsideABlockThenDeleteFromRest implements ISolvingAlgorithm {
	// If potential values are only possible on one line/column of a block (as others are fixed, for example)
	//     then delete this potential value from the rest
	@Override
	public boolean tryToSolve(Grid grid) {
		// check all blocks
		for (Block block: grid.getBlocks()) {
			// check all cells inside the block
			for (Cell cell: block.getCells()) {
				// check all potential values inside the cell
				for (OneToNine potentialValue: cell.getPotentialValues()) {
					if (checkIfValueIsOnlyInsideTheCellsLine(block, cell, potentialValue)) {
						deleteValueInRestOutsideBlock(block, cell, potentialValue, cell.getLine().getCells());
					}
					if (checkIfValueIsOnlyInsideTheCellsColumn(block, cell, potentialValue)) {
						deleteValueInRestOutsideBlock(block, cell, potentialValue, cell.getColumn().getCells());
					}
				}
			}
		}

		return grid.isSolved();
	}

	// delete the potential value in all cells inside its line/column (but outside of its block)
	private void deleteValueInRestOutsideBlock(Block block, Cell cell, OneToNine potentialValue, List<Cell> restOfCells) {
		Block theCellsBlock = cell.getBlock();
		if (theCellsBlock != block) throw new RuntimeException("should never happen"); // paranoia

		for (Cell otherCell: restOfCells) {
			if (!theCellsBlock.contains(otherCell)) {
				otherCell.deleteValue(potentialValue);
			}
		}
	}

	private boolean checkIfValueIsOnlyInsideTheCellsColumn(Block block, Cell cell, OneToNine value) {
		// check all cells in the block
		for (Cell otherCellInBlock: block.getCells()) {
			// only look at cells which are inside the same block but are in another line
			if (cell.getColumn() != otherCellInBlock.getColumn()) {
				// if one of these cells contains the potential value , then we are done
				if (otherCellInBlock.getPotentialValues().contains(value)) {
					return false;
				}
			}
		}

		return true;
	}

	private boolean checkIfValueIsOnlyInsideTheCellsLine(Block block, Cell cell, OneToNine value) {
		// check all cells in the block
		for (Cell otherCellInBlock: block.getCells()) {
			// only look at cells which are inside the same block but are in another line
			if (cell.getLine() != otherCellInBlock.getLine()) {
				// if one of these cells contains the potential value , then we are done
				if (otherCellInBlock.getPotentialValues().contains(value)) {
					return false;
				}
			}
		}

		return true;
	}
}

File: workspace/scudoku/src/main/scala/scudoku/solvers/IfValueFixedToLineOrColumnInsideABlockThenDeleteFromRest.scala
package scudoku.solvers

import scudoku.data.Grid
import scudoku.data.UnknownOrOneToNine
import scudoku.data.UnknownOrOneToNine._
import scudoku.data.Cell
import scudoku.data.Block
import scudoku.data.CellSet

class IfValueFixedToLineOrColumnInsideABlockThenDeleteFromRest extends TSolvingAlgorithm {
  override def tryToSolve(grid: Grid): Boolean = {
    // check all blocks
    grid.blocks.
      map(block => block.getCells).
      // check all cells inside the block
      foreach(cells => checkAllPotentialValuesInsideCells(cells))
     
    grid.isSolved
  }
  
  private def checkAllPotentialValuesInsideCells(cells: Traversable[Cell]) = {
    cells.
    foreach(cell => {
      checkAllPotentialValuesInsideTheCellsLineOrColumn(cell, (cell => cell.line))
      checkAllPotentialValuesInsideTheCellsLineOrColumn(cell, (cell => cell.column))
    })
  }
  
  private def checkAllPotentialValuesInsideTheCellsLineOrColumn(cell: Cell, mapCellToLineOrColumn: (Cell => CellSet)) = {
    // check all potential values inside the cell ... for line or column
    cell.potentialValues.
      filter(potentialValue => checkIfPotentialValueIsOnlyInsideTheCellsLineOrColumn(cell, potentialValue, mapCellToLineOrColumn)).
      foreach(potentialValue => deleteValueInRestOutsideBlock(cell, potentialValue, mapCellToLineOrColumn(cell).getCells))
  }
  
  private def checkIfPotentialValueIsOnlyInsideTheCellsLineOrColumn(cell: Cell, value: UnknownOrOneToNine, mapCellToLineOrColumn: (Cell => CellSet)): Boolean = {
    // check all cells in the block
    ! cell.block.getCells.
       // only look at cells which are inside the same block but are in another column
      filter(otherCellInBlock => (mapCellToLineOrColumn(cell) != mapCellToLineOrColumn(otherCellInBlock))).
      // if one of these cells contains the potential value, then we are done
      exists(otherCellInBlock => otherCellInBlock.potentialValues.contains(value))
  }
  
  // delete the potential value in all cells inside its line/column (but outside of its block)
  private def deleteValueInRestOutsideBlock(cell: Cell, potentialValue: UnknownOrOneToNine, restOfCells: Traversable[Cell]) = {
    restOfCells.
      filter(otherCell => !cell.block.contains(otherCell)).
      foreach(otherCell => otherCell.deletePotentialValue(potentialValue))
  }
}